dask_sql/physical/utils/sort.py

```diff
@@ -6,12 +6,34 @@
 from dask_sql.utils import make_pickable_without_dask_sql, new_temporary_column
 
 
+def multi_col_sort(
+    df: dd.DataFrame,
+    sort_columns: List[str],
+    sort_ascending: List[bool],
+    sort_null_first: List[bool],
+) -> dd.DataFrame:
+
+    df = df.sort_values(sort_columns)
+    return df.persist()
+
+
 def apply_sort(
     df: dd.DataFrame,
     sort_columns: List[str],
     sort_ascending: List[bool],
     sort_null_first: List[bool],
 ) -> dd.DataFrame:
+
+    # Try fast path for multi-column sorting before falling back to
+    # sort_partition_func.  Tools like dask-cudf have a limited but fast
+    # multi-column sort implementation.  We check if any sorting/null sorting
+    # is required.  If so, we fall back to default sorting implementation
+    if any(sort_null_first) is False and all(sort_ascending) is True:
```

Comment:
So by default, the multi column sort in dask-cudf sorts nulls last and sorts everything ascending? (I guess so, just wanted to double check)

Reply:
Because (and that might be actually a problem), the `dask.DataFrame.sort_values` function does some "strange" (at least from the point of SQL) things to NaN values. As it is basically computing new divisions, which can not contain NaN values, it will just remove all NaN rows. You can test this:
```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame({"a": list(range(200)) + [None]*100})
ddf = dd.from_pandas(df, npartitions=11)

ddf.sort_values('a')
```

will return a Dataframe with length 200 (instead of 300). Even if the user did not specify to sort NaN values in a non-default way, I think we should not have this function as default - for Dask. Is this different in Dask-cudf (or cudf)? I am fine to have a special case handling for dask-cudf if it makes sense.

We actually should have a test for this. I just pushed one (dirty without PR, ehem) onto main in 9ebe16a53e7bbb89314061b38014e03f663e6a5b. This works on main, but fails with your PR (due to the issue I described).

Reply:
Thanks for pointing this out! I think the additional condition `isinstance(df, dask_cudf.DataFrame)` might be necessary here to avoid Dask's single column case.

To answer your question, dask-cudf doesn't implement null sorting:

```python
import dask_cudf

gddf = dask_cudf.from_dask_dataframe(ddf)
gddf.sort_values("a")
```

Results in

```
---------------------------------------------------------------------------                                        
ValueError                                Traceback (most recent call last)                                        
<ipython-input-7-6fc4356cae9b> in <module>                                                                         
----> 1 gddf.sort_values("a")                                                                                      
                                                                                                                   
/raid/charlesb/miniconda3/envs/dask/lib/python3.8/site-packages/dask_cudf/core.py in sort_values(self, by, ignore_i
ndex, max_branch, divisions, set_divisions, **kwargs)                                                              
    244             df = self.map_partitions(M.sort_values, by)                                                    
    245         else:                                                                                              
--> 246             df = sorting.sort_values(                                                                      
    247                 self,                                                                                      
    248                 by,                                                                                        
                                                                                                                   
/raid/charlesb/miniconda3/envs/dask/lib/python3.8/site-packages/dask_cudf/sorting.py in sort_values(df, by, max_bra
nch, divisions, set_divisions, ignore_index)                                                                       
    223     # Step 1 - Calculate new divisions (if necessary)                                                      
    224     if divisions is None:                                                                                  
--> 225         divisions = quantile_divisions(df, by, npartitions)                                                
    226                                                                                                            
    227     # Step 2 - Perform repartitioning shuffle                                                              
                                                                                                                   
/raid/charlesb/miniconda3/envs/dask/lib/python3.8/site-packages/dask_cudf/sorting.py in quantile_divisions(df, by, 
npartitions)                                                                                                       
    188         divisions.iloc[-1] += 1                                                                            
    189         divisions = sorted(                                                                                
--> 190             divisions.drop_duplicates().astype(dtype).values.tolist()                                      
    191         )                                                                                                  
    192     else:                                                                                                  
                                                                                                                   
/raid/charlesb/miniconda3/envs/dask/lib/python3.8/site-packages/cudf/core/frame.py in values(self)                 
   4528         <class 'cupy.core.core.ndarray'>                                                                   
   4529         """                                                                                                
-> 4530         return self._column.values                                                                         
   4531                                                                                                            
   4532     @property                                                                                              
                                                                                                                   
/raid/charlesb/miniconda3/envs/dask/lib/python3.8/site-packages/cudf/core/column/column.py in values(self)         
    142                                                                                                            
    143         if self.has_nulls:                                                                                 
--> 144             raise ValueError("Column must have no nulls.")                                                 
    145                                                                                                            
    146         return cupy.asarray(self.data_array_view)                                                          
                                                                                                                   
ValueError: Column must have no nulls. 
```

With this in mind, we should probably adjust our try/except block to look out for errors on dask-cudf's end that can be handled by dask-sql (i.e. presence of nulls in the dataframe).